### PR TITLE
use GenK for testdata generation

### PR DESCRIPTION
**Pull request by abendt**
use GenK to generate testdata. this is a part of (#1858)

modified laws are:
* BicrosswalkLaws
* BifoldableLaws
* BitraverseLaws

### CODE REVIEW

**Comment by abendt** [comment]
one question regarding the use of GenK/EqK for Kind2 types. what i like about the tests using unary type constructors the actual tests do not need any knowledge about the actual inner type.
in constrast the tests here always need to know of one of the inner types.
wouldn't it be better to use something like Gen2K/Eq2K here?

**Comment by 1Jajen1** [comment]
Good idea :+1:. With arity 2 we should have basically every typeclass covered.

**Comment by abendt** [comment]
> Good idea 👍. With arity 2 we should have basically every typeclass covered.

i have prepared a quick spike so you can see how it would look like: https://github.com/abendt/arrow/tree/gen2k

seems to be a straightforward change

**Comment by abendt** [comment]
i changed the laws to use Gen2K/Eq2K now

**Comment by abendt** [comment]
i have merged in my other PR with CategoryLaws and ProfunctorLaws and changed the laws to use Gen2K/Eq2K too

File: modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Eq2K.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package arrow.typeclasses
+
+import arrow.Kind2
+
+/**
+ * The `Eq2K` typeclass abstracts the ability to lift the Eq class to binary type constructors.
+ */
+interface Eq2K<F> {
```
Comment (1Jajen1) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Eq2K.kt]: I think `EqK2` is a better name as I would read it as "EqualityKind2".

File: modules/core/arrow-core-data/src/test/kotlin/arrow/core/AndThenTest.kt [hunk 2]
```diff
@@ -119,9 +122,24 @@ private fun <A> AndThen.Companion.eqK() = object : EqK<AndThenPartialOf<A>> {
     }
 }
 
+private fun AndThen.Companion.eq2K() = object : Eq2K<ForAndThen> {
+  override fun <A, B> Kind2<ForAndThen, A, B>.eqK(other: Kind2<ForAndThen, A, B>, EQA: Eq<A>, EQB: Eq<B>): Boolean =
+    (this.fix() to other.fix()).let { (ls, rs) ->
+      EQB.run {
+        ls(1).eqv(rs(1))
```
Comment (1Jajen1) [inline modules/core/arrow-core-data/src/test/kotlin/arrow/core/AndThenTest.kt]: How does this even typecheck? o.O Shouldn't the input be `A` and not `Int`?

File: modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eq2k/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+## Eq2K
+
+The `Eq2K` typeclass abstracts the ability to lift the Eq class to binary type constructors.
+
+### Main Combinators
+
+#### Kind2<F, A, B>.eqK(other: Kind2<F, A, B>, EQA: Eq<A>, EQB: Eq<B>): Boolean
+
+Compares two instances of a type with binary typeconstructor in the context of `F` using the provided `Eq<A>` and `Eq<B>`. Returns true if they're considered equal.
```
Comment (1Jajen1) [inline modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eq2k/README.md]:
```suggestion
Compares two instances of a type `F` with binary a type constructor using the provided `Eq<A>` and `Eq<B>`. Returns true if they're considered equal.
```

File: modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eq2k/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+// lift String Eq to Either
+Either.eq2K().run {
+    Either.right("hello").eqK(Either.right("arrow"), String.eq(), String.eq())
+}
+```
+
+### Laws
+
+Arrow provides `Eq2KLaws` in the form of test cases for internal verification of lawful instances and third party apps creating their own `Eq2K` instances.
+See the existing EqK instances implementations and accompanying tests for reference.
```
Comment (1Jajen1) [inline modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eq2k/README.md]:
```suggestion
See the existing EqK2 instances implementations and accompanying tests for reference.
```
Change this to `Eq2K` if we keep that name

File: modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Eq2K.kt (hunk 1 above)
Comment (abendt) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Eq2K.kt]: yes makes sense. i have changed it

File: modules/core/arrow-core-data/src/test/kotlin/arrow/core/AndThenTest.kt (hunk 2 above)
Comment (abendt) [reply modules/core/arrow-core-data/src/test/kotlin/arrow/core/AndThenTest.kt]: i just adapted the existing code 😁 isn't 'A' just resolved to 'Int' here?

File: modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eqk2/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+## EqK2
+
+The `EqK2` typeclass abstracts the ability to lift the Eq class to binary type constructors.
+
+### Main Combinators
+
+#### Kind2<F, A, B>.eqK(other: Kind2<F, A, B>, EQA: Eq<A>, EQB: Eq<B>): Boolean
+
+Compares two instances of a type `F` with binary a type constructor using the provided `Eq<A>` and `Eq<B>`. Returns true if they're considered equal.
```
Comment (1Jajen1) [inline modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eqk2/README.md]:
```suggestion
Compares two instances of a type `F` with a binary type constructor using the provided `Eq<A>` and `Eq<B>`. Returns true if they're considered equal.
```
My bad :man_facepalming:

Comment (1Jajen1) [inline modules/docs/arrow-docs/docs/docs/arrow-core/typeclasses/eqk2/README.md]:
```suggestion
Compares two instances of a type `F` with a binary type constructor using the provided `Eq<A>` and `Eq<B>`. Returns true if they're considered equal.
```
My bad :man_facepalming:

**Pull request merged**